src/main/kotlin/com/openlattice/transporter/TransporterQueries.kt

```diff
@@ -0,0 +1,145 @@
+    val destCol = propCol.destColName
+    val dataCol = "CASE WHEN VERSION > 0 THEN $srcCol else null END as $destCol"
+
+    val updateLastPropagate = "UPDATE ${PostgresTable.DATA.name} " +
+            "SET ${transportTimestampColumn.name} = ${DataTables.LAST_WRITE.name} " +
+            "WHERE ${PARTITION.name} = ANY(?) " +
+            " AND ${ENTITY_SET_ID.name} = ANY(?) " +
+            " AND ${PROPERTY_TYPE_ID.name} = '${propId}' " +
+            " AND ${DataTables.LAST_WRITE.name} > ${transportTimestampColumn.name} " +
+            "RETURNING $ids,${VERSION.name},$dataCol"
+
+    // this seems unnecessary but it's good to be sure
+    val createMissingRows = "INSERT INTO $destTable " +
+            "($ids) " +
+            "SELECT $ids " +
+            "FROM src " +
+            "WHERE ${VERSION.name} > 0 " +
+            "ON CONFLICT ($pk) DO NOTHING"
+    val modifyDestination = "UPDATE $destTable t " +
+            "SET $destCol = src.$destCol " +
+            "FROM src " +
+            "WHERE " + pkCols.joinToString(" AND ") { "t." + it.name + " = src." + it.name}
+
+    return "WITH src as (${updateLastPropagate}), inserted as ($createMissingRows) $modifyDestination"
+}
+
+/**
+ * column bindings are
+ * 1 - partitions array
+ * 2 - entity set ids array
+ *
+ */
+fun updateIdsForEntitySets(destTable: String): String {
+    val updateLastPropagate = "UPDATE ${PostgresTable.IDS.name} " +
+            "SET ${transportTimestampColumn.name} = ${DataTables.LAST_WRITE.name} " +
+            "WHERE ${PARTITION.name} = ANY(?) " +
+            " AND ${ENTITY_SET_ID.name} = ANY(?) " +
+            " AND ${DataTables.LAST_WRITE.name} > ${transportTimestampColumn.name} " +
+            "RETURNING ${ENTITY_SET_ID.name}, ${ID_VALUE.name},${LINKING_ID.name},${VERSION.name}"
+    val createMissingRows = "INSERT INTO $destTable ($pk) " +
+            "SELECT $pk " +
+            "FROM src " +
+            "WHERE ${VERSION.name} > 0 " +
+            "ON CONFLICT ($pk) DO NOTHING"
+    val createMissingLinkedRows = "INSERT INTO $destTable ($pk,${ORIGIN_ID.name}) " +
+            "SELECT ${ENTITY_SET_ID.name}, " +
+            " ${LINKING_ID.name} AS ${ID_VALUE.name}, " +
+            " ${ID_VALUE.name} AS ${ORIGIN_ID.name} " +
+            "FROM src " +
+            "WHERE ${VERSION.name} > 0 " +
+            " AND ${LINKING_ID.name} IS NOT NULL " +
+            "ON CONFLICT ($pk) DO NOTHING"
+    val deleteRows = "DELETE FROM $destTable t " +
+            "USING src " +
+            "WHERE src.${VERSION.name} <= 0 " +
+            " AND t.${ENTITY_SET_ID.name} = src.${ENTITY_SET_ID.name} " +
+            " AND t.${ID_VALUE.name} in (src.${ID_VALUE.name},src.${LINKING_ID.name})"
+    return "WITH src as ($updateLastPropagate), " +
+            "inserts as ($createMissingRows)," +
+            "insertLinks as ($createMissingLinkedRows) " +
+            deleteRows
+}
+
+fun addAllMissingColumnsQuery(table: PostgresTableDefinition): String =
```

Comment:
unclear if we are about this right now, but property types can also be removed from entity types -- are we going to remove those columns in that case? (or just leave that data in case it's re-added and remove it from the MVs)

Reply:
We should remove the columns eventually, yes. EdmService only allows this when there aren't any entity sets for a data type, so there should never be stale data in the column (but there could be an all-nulls column hanging out forever).